core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,271 @@
+                flush()
+                seenErrors[errorKey] = 0 // Zero because it's already been sent.
+            }
+        }
+        else {
+            addRemoteMetric(metric, filteredTags, log=logData)
+        }
+    }
+
+    @Synchronized
+    fun flush() {
+        if (!enable || queue.isEmpty()) return
+        if (rateLimitEndTime > (System.currentTimeMillis() / 1000).toInt()) {
+            return
+        }
+        rateLimitEndTime = 0
+
+        try {
+            send()
+            queueBytes = 0
+        } catch (error: Throwable) {
+            errorHandler?.invoke(error)
+            sampleRate = 0.0
+        }
+    }
+
+    private fun send() {
+        try {
+            // Json.encodeToString by default does not include default values
+            //  We're using this to leave off the 'log' parameter if unset.
+            val payload = Json.encodeToString(mapOf("series" to queue))
+            resetQueue()
+
+            val connection = httpClient.upload(host)
+            connection.outputStream?.use { outputStream ->
+                // Write the JSON string to the outputStream.
+                outputStream.write(payload.toByteArray(Charsets.UTF_8))
+                outputStream.flush() // Ensure all data is written
+            }
+            connection.close()
+        } catch (e: HTTPException) {
+            errorHandler?.invoke(e)
+            if (e.responseCode == 429) {
+                val headers = e.responseHeaders
+                val rateLimit = headers["Retry-After"]?.firstOrNull()?.toLongOrNull()
+                if (rateLimit != null) {
+                    rateLimitEndTime = rateLimit + (System.currentTimeMillis() / 1000)
+                }
+            }
+        } catch (e: Exception) {
+            errorHandler?.invoke(e)
+        }
+    }
+
+    private val additionalTags: Map<String, String> by lazy {
+        var osVersion = System.getProperty("os.version")
+        val androidRegex = Regex("android[0-9][0-9]")
+        val androidMatch = androidRegex.find(osVersion)
+        if (androidMatch != null) {
+            osVersion = androidMatch.value
+        } else {
+            // other OSes seem to have well formed version numbers, just grab first for major version
+            val otherRegex = Regex("[0-9]+")
+            val otherMatch = otherRegex.find(osVersion)
+            if (otherMatch != null) {
+                osVersion = otherMatch.value
+            }
+        }
+        mapOf(
+            "os" to System.getProperty("os.name") + "-" + osVersion,
+            "interpreter" to System.getProperty("java.vendor") + "-" + System.getProperty("java.version"),
+            "library" to "analytics.kotlin",
+            "library_version" to Constants.LIBRARY_VERSION
+        )
+    }
+
+    private fun addRemoteMetric(metric: String, tags: Map<String, String>, value: Int = 1, log: String? = null) {
+        val found = queue.find {
```

Comment:
this might be too slow. we should use a map for quick look up

Reply:
Could you outline what you're thinking for this?